## 1 네이버 뉴스 수집

In [1]:
import requests
from bs4 import BeautifulSoup

url = "https://news.naver.com/main/read.nhn?mode=LSD&mid=sec&sid1=105&oid=030&aid=0002700985"

res = requests.get(url)
res

<Response [503]>

503 error가 발생하였는데, 503은 Service not available

url에 접속해서 개발자도구 -> network에 가보면 request가 존재하는데, header -> request 부분을 보면 user-agent가 있음

이 정보를 명시해야 server에서 어떤 브라우저를 통해 요청한지 알 수 있음

In [2]:
import requests
from bs4 import BeautifulSoup

url = "https://news.naver.com/main/read.nhn?mode=LSD&mid=sec&sid1=105&oid=030&aid=0002700985"
header = {'user-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.4664.110 Safari/537.36'}

res = requests.get(url, headers = header)
res

<Response [200]>

header를 추가하게되면, 200을 받게되고 정상적으로 처리된걸 알 수 있음

In [3]:
# 수신한 데이터 확인
res.content

b'<!DOCTYPE HTML> \r\n<html lang="ko"> \r\n<head>\r\n<meta charset="euc-kr">\r\n<meta http-equiv="X-UA-Compatible" content="IE=edge">\r\n<meta name="referrer" contents="always">\r\n<meta name="viewport" content="width=1106" />\r\n<title>\xbc\xf6\xbf\xf8\xbd\xc3 \xb0\xf8\xc0\xaf \xc0\xda\xc0\xfc\xb0\xc5, \xbb\xf3\xb9\xdd\xb1\xe2 4050\xb4\xeb \xc3\xdf\xb0\xa1 : \xb3\xd7\xc0\xcc\xb9\xf6 \xb4\xba\xbd\xba</title>\r\n\n\n\t\n\t\n\t\t\n\t\n\n\n\t\n\t\n\t\n\t\t\n\t\t\n\t\t\n\t\n\n<meta property="me2:post_tag"\t\tcontent="\xc0\xfc\xc0\xda\xbd\xc5\xb9\xae : \xb3\xd7\xc0\xcc\xb9\xf6\xb4\xba\xbd\xba"/>\n<meta property="me2:category1"\t\tcontent="\xc0\xfc\xc0\xda\xbd\xc5\xb9\xae"/>\n<meta property="me2:category2"\t\tcontent="IT/\xb0\xfa\xc7\xd0"/>\n<meta property="me2:image"\t\t\tcontent="https://imgnews.pstatic.net/image/030/2018/04/20/0002700985_001_20180420234703177.jpg"/>\n\n<meta property="og:title"\t\t\tcontent="\xbc\xf6\xbf\xf8\xbd\xc3 \xb0\xf8\xc0\xaf \xc0\xda\xc0\xfc\xb0\xc5, \xbb\xf3\xb9\

In [4]:
# beautifulSoup를 통해 수신한 데이터에서 데이터를 추출
bs = BeautifulSoup(res.content, 'html.parser') # res.content에서 html.parser로 파싱하여 bs에 저장

# bs에서 데이터 뽑기
# url 페이지에서 개발자도구 -> Elements -> 왼쪽 상단 마우스 포인터 클릭 -> 뉴스 기사의 날짜 클릭 -> span 태그 밑에 t11 클래스에 날짜가 들어있는것을 확인 가능
bs.select('span.t11')

[<span class="t11">2018.04.20. 오후 11:47</span>]

In [5]:
# 날짜 데이터만 뽑기
date = bs.select('span.t11')[0].text # 리스트 형태로 오는데, 데이터 0 번째에서 text만 추출
date

'2018.04.20. 오후 11:47'

In [6]:
# 제목 가져오기 
title = bs.select('h3#articleTitle')[0].text # id의 경우 #을 써서 사용
title

'수원시 공유 자전거, 상반기 4050대 추가'

In [7]:
# 매체 명 가져오기
media = bs.select('div.press_logo a img')[0]['title']
media

'전자신문'

In [8]:
# 본문 가져오기
content = bs.select('div#articleBodyContents')[0].get_text().replace('\n', '').replace('\t', '')
content

"// flash 오류를 우회하기 위한 함수 추가function _flash_removeCallback() {}수원시가 지난해 도입한 '스테이션 없는 무인대여자전거' 숫자가 대폭 늘어난다.'모바이크'는 상반기 안에 4050대를 추가한다. 5월 중 1400대, 6월 중 2650대를 배치할 계획이다. 모바이크는 수원시에서 무인대여자전거 950대를 운영하는 공유자전거 업체다.새롭게 배치되는 자전거는 '뉴라이트 모델'로 기존 자전거보다 성능이 개선됐다. 자전거 무게를 6㎏(22㎏→16㎏) 줄였고, 잠금장치 배터리 방전을 방지하기 위한 태양광 패널을 부착했다. 여성·아동도 쉽게 탈 수 있도록 안장 높이 조절 범위를 늘렸다.수원시에 도입되는 모바이크 공유자전거.회원 가입할 때 내야 했던 보증금(5000원)은 23일부터 없앤다. 국내 모든 신용카드로 이용료를 결제할 수 있도록 결제 시스템을 개선한다.수원시는 5월부터 무인대여 자전거 주차구역 관리(자전거 정리정돈·거치대 청소·방치자전거 처리 등)를 민간기관에 위탁해 시민 불편을 최소화한다. 자전거 이용경로를 알려주는 빅데이터를 활용해 자전거 주차장을 확보하고, 자전거도로 정비 우선순위를 정할 예정이다.무인대여자전거 시스템은 자전거 거치대, 무인 정보안내시스템 등이 필요 없다. 스마트폰을 활용해 GPS가 장착된 자전거를 수원시 곳곳에 있는 자전거 주차공간에서 간편하게 대여하고 반납할 수 있다. 수원시에서는 공유자전거 업체인 모바이크와 오바이크가 무인대여 자전거를 운영한다.무인대여 자전거를 이용하려면 먼저 스마트폰 플레이스토어(안드로이드폰), 앱스토어(아이폰)에서 오바이크나 모바이크를 검색해 애플리케이션(앱)을 내려받아야 한다.무인자전거 앱을 활성화하면 주변 자전거 위치를 알려주는 지도가 나온다. 자전거가 있는 장소를 찾아가 앱 하단 '잠금 해제'를 누른 후 스마트폰으로 자전거 핸들 사이에 부착된 QR코드를 스캔하면 잠금이 해제된다.자전거를 타고 목적지까지 이동한 후 목적지 주변에 있는 '공공자전거 주차공간'(노면에 표시)이나

In [9]:
# 본문 가져오기 2
# flash오류를 막기위해 공통적으로 들어간 첫 줄 제거
content = bs.select('div#articleBodyContents')[0].get_text().replace('\n', '').replace('\t', '')
content = content.replace('// flash 오류를 우회하기 위한 함수 추가function _flash_removeCallback() {}', '')
content

"수원시가 지난해 도입한 '스테이션 없는 무인대여자전거' 숫자가 대폭 늘어난다.'모바이크'는 상반기 안에 4050대를 추가한다. 5월 중 1400대, 6월 중 2650대를 배치할 계획이다. 모바이크는 수원시에서 무인대여자전거 950대를 운영하는 공유자전거 업체다.새롭게 배치되는 자전거는 '뉴라이트 모델'로 기존 자전거보다 성능이 개선됐다. 자전거 무게를 6㎏(22㎏→16㎏) 줄였고, 잠금장치 배터리 방전을 방지하기 위한 태양광 패널을 부착했다. 여성·아동도 쉽게 탈 수 있도록 안장 높이 조절 범위를 늘렸다.수원시에 도입되는 모바이크 공유자전거.회원 가입할 때 내야 했던 보증금(5000원)은 23일부터 없앤다. 국내 모든 신용카드로 이용료를 결제할 수 있도록 결제 시스템을 개선한다.수원시는 5월부터 무인대여 자전거 주차구역 관리(자전거 정리정돈·거치대 청소·방치자전거 처리 등)를 민간기관에 위탁해 시민 불편을 최소화한다. 자전거 이용경로를 알려주는 빅데이터를 활용해 자전거 주차장을 확보하고, 자전거도로 정비 우선순위를 정할 예정이다.무인대여자전거 시스템은 자전거 거치대, 무인 정보안내시스템 등이 필요 없다. 스마트폰을 활용해 GPS가 장착된 자전거를 수원시 곳곳에 있는 자전거 주차공간에서 간편하게 대여하고 반납할 수 있다. 수원시에서는 공유자전거 업체인 모바이크와 오바이크가 무인대여 자전거를 운영한다.무인대여 자전거를 이용하려면 먼저 스마트폰 플레이스토어(안드로이드폰), 앱스토어(아이폰)에서 오바이크나 모바이크를 검색해 애플리케이션(앱)을 내려받아야 한다.무인자전거 앱을 활성화하면 주변 자전거 위치를 알려주는 지도가 나온다. 자전거가 있는 장소를 찾아가 앱 하단 '잠금 해제'를 누른 후 스마트폰으로 자전거 핸들 사이에 부착된 QR코드를 스캔하면 잠금이 해제된다.자전거를 타고 목적지까지 이동한 후 목적지 주변에 있는 '공공자전거 주차공간'(노면에 표시)이나 자전거 거치대에 세운 후 뒷바퀴 윗부분에 있는 잠금장치를 채워놓으면 된다. 이용료는 모바이크가 30분에 30

In [11]:
# 함수화
def get_naver_news(url):
  header = {'user-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.4664.110 Safari/537.36'}

  res = requests.get(url, headers = header)
  bs = BeautifulSoup(res.content, 'html.parser')

  date = bs.select('span.t11')[0].text 
  title = bs.select('h3#articleTitle')[0]
  media = bs.select('div.press_logo a img')[0]['title']
  content = bs.select('div#articleBodyContents')[0].get_text().replace('\n', '').replace('\t', '')
  content = content.replace('// flash 오류를 우회하기 위한 함수 추가function _flash_removeCallback() {}', '')

  return date, title, media, content

In [12]:
url = "https://news.naver.com/main/read.nhn?mode=LSD&mid=sec&sid1=105&oid=030&aid=0002700985"

get_naver_news(url)

('2018.04.20. 오후 11:47',
 <h3 id="articleTitle">수원시 공유 자전거, 상반기 4050대 추가</h3>,
 '전자신문',
 "수원시가 지난해 도입한 '스테이션 없는 무인대여자전거' 숫자가 대폭 늘어난다.'모바이크'는 상반기 안에 4050대를 추가한다. 5월 중 1400대, 6월 중 2650대를 배치할 계획이다. 모바이크는 수원시에서 무인대여자전거 950대를 운영하는 공유자전거 업체다.새롭게 배치되는 자전거는 '뉴라이트 모델'로 기존 자전거보다 성능이 개선됐다. 자전거 무게를 6㎏(22㎏→16㎏) 줄였고, 잠금장치 배터리 방전을 방지하기 위한 태양광 패널을 부착했다. 여성·아동도 쉽게 탈 수 있도록 안장 높이 조절 범위를 늘렸다.수원시에 도입되는 모바이크 공유자전거.회원 가입할 때 내야 했던 보증금(5000원)은 23일부터 없앤다. 국내 모든 신용카드로 이용료를 결제할 수 있도록 결제 시스템을 개선한다.수원시는 5월부터 무인대여 자전거 주차구역 관리(자전거 정리정돈·거치대 청소·방치자전거 처리 등)를 민간기관에 위탁해 시민 불편을 최소화한다. 자전거 이용경로를 알려주는 빅데이터를 활용해 자전거 주차장을 확보하고, 자전거도로 정비 우선순위를 정할 예정이다.무인대여자전거 시스템은 자전거 거치대, 무인 정보안내시스템 등이 필요 없다. 스마트폰을 활용해 GPS가 장착된 자전거를 수원시 곳곳에 있는 자전거 주차공간에서 간편하게 대여하고 반납할 수 있다. 수원시에서는 공유자전거 업체인 모바이크와 오바이크가 무인대여 자전거를 운영한다.무인대여 자전거를 이용하려면 먼저 스마트폰 플레이스토어(안드로이드폰), 앱스토어(아이폰)에서 오바이크나 모바이크를 검색해 애플리케이션(앱)을 내려받아야 한다.무인자전거 앱을 활성화하면 주변 자전거 위치를 알려주는 지도가 나온다. 자전거가 있는 장소를 찾아가 앱 하단 '잠금 해제'를 누른 후 스마트폰으로 자전거 핸들 사이에 부착된 QR코드를 스캔하면 잠금이 해제된다.자전거를 타고 목적지까지 이동한 후 목적지 주

# 2 네이버 블로그 데이터 수집

In [43]:
import requests
from bs4 import BeautifulSoup

url = 'https://blog.naver.com/biomath2k/221755824489'

header = {'user-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.4664.110 Safari/537.36'}

res = requests.get(url, headers = header)
bs = BeautifulSoup(res.content, 'html.parser')

In [44]:
# 날짜 가져오기
bs.select('span.se_publishDate')

[]

데이터가 없다고 나오는 이유는, 최상위에 iframe(inner frame)으로 되어있기 때문
그러므로 mainFrame에서 url로 한번 더 이동한것을 따라가서 데이터를 가져와야 함

In [45]:
new_url = 'https://blog.naver.com/' + bs.select('iframe')[0]['src'] # iframe에서 주소 가져와서 url을 변경
res = requests.get(new_url, headers = header)
bs = BeautifulSoup(res.content, 'html.parser')

In [46]:
# 날짜 가져오기
date = bs.select('span.se_publishDate')[0].text
date

'2019. 12. 31. 23:50'

In [47]:
title = bs.select('div.se-title-text p span')[0].text.replace('\n', '')
title

'[IT] 삼성 청년 SW 아카데미 (SSAFY)'

In [49]:
content = bs.select('div.se-main-container')[0].get_text().replace('\n', '').replace('\u200b', '')
content

'삼성 청년 SW 아카데미는 삼성전자가 주관하고 고용노동부가 후원하는, 청년들의 취업 성공을 돕는 프로그램입니다.   삼성전자가 2019년 11월18일 서울 멀티캠퍼스 교육센터를 포함해 대전, 광주, 구미 등 4개 교육센터에서 삼성 청년 소프트웨어 아카데미 SSAFY(Samsung Software Academy For Youth) 1기 수료식을 열었다.‘삼성 청년 소프트웨어 아카데미’는 청소년 교육을 테마로 한 삼성전자  사회공헌 프로그램 중 하나로, 소프트웨어 경쟁력 강화를 통해 국내 IT 생태계 저변을 확대하고 청년 취업 경쟁력을 높이기 위해 미취업 청년들을 대상으로 실시하는 소프트웨어 교육 프로그램이다.SSAFY 제현웅 상무는 “오늘 수료하는 학생들은 모두 1년간 매일 8시간씩의 강도 높은 교육을 성실히 이수했으며 열정과 끈기를 겸비했다”라며 “삼성 청년 소프트웨어 아카데미 출신 인력들이 그동안 쌓은 역량을 발휘하며 여러 기업에서 활약하기를 기대한다”라고 말했다.삼성전자는 2018년 12월 고용노동부와 ‘삼성 청년 소프트웨어 아카데미 운영지원을 위한 업무협약(MOU)’을 체결해 교육생 선발을 위한 설명회 실시, 교육과정 홍보, 양질의 소프트웨어 교육 등 전반적인 운영부터 취업활동 지원까지 다방면으로 긴밀하게 협력하고 있다.1년간 1∙2학기 과정으로 나눠, 1학기는 기본기를 다지는 코딩 기본 교육, 2학기는 실무 중심의 프로젝트 수행 교육으로 진행되며 교육생 전원은 매달 100만원의 지원비를 받는다.1학기 기본과정은 5개월 동안 알고리즘과 코딩의 소프트웨어 기본기를 다지는 과정으로, 개인 역량에 따라 분반해 파이썬, 자바 등 프로그래밍 언어와 알고리즘, 프레임워크, 데이터베이스를 중점 학습한다.2학기 심화과정은 인공지능(AI), 사물인터넷(IoT) 등 4차 산업혁명 기술을 활용한 프로젝트를 통해 소프트웨어 실전역량을 기르는 과정으로 5개월 동안 진행된다. 1학기와 달리 교재나 별도의 이론 강의 없이 100%프로젝트 기반으로 진행되며 자기주도형 학습을 통해 

In [50]:
def get_naver_blog(url):
  header = {'user-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.4664.110 Safari/537.36'}

  res = requests.get(url, headers = header)
  bs = BeautifulSoup(res.content, 'html.parser')
  
  new_url = 'https://blog.naver.com/' + bs.select('iframe')[0]['src']
  res = requests.get(new_url, headers = header)
  bs = BeautifulSoup(res.content, 'html.parser')

  date = bs.select('span.se_publishDate')[0].text
  title = bs.select('div.se-title-text p span')[0].text.replace('\n', '')
  content = bs.select('div.se-main-container')[0].get_text().replace('\n', '').replace('\u200b', '')

  return date, title, content

In [51]:
url = 'https://blog.naver.com/biomath2k/221755824489'

get_naver_blog(url)

('2019. 12. 31. 23:50',
 '[IT] 삼성 청년 SW 아카데미 (SSAFY)',
 '삼성 청년 SW 아카데미는 삼성전자가 주관하고 고용노동부가 후원하는, 청년들의 취업 성공을 돕는 프로그램입니다.   삼성전자가 2019년 11월18일 서울 멀티캠퍼스 교육센터를 포함해 대전, 광주, 구미 등 4개 교육센터에서 삼성 청년 소프트웨어 아카데미 SSAFY(Samsung Software Academy For Youth) 1기 수료식을 열었다.‘삼성 청년 소프트웨어 아카데미’는 청소년 교육을 테마로 한 삼성전자  사회공헌 프로그램 중 하나로, 소프트웨어 경쟁력 강화를 통해 국내 IT 생태계 저변을 확대하고 청년 취업 경쟁력을 높이기 위해 미취업 청년들을 대상으로 실시하는 소프트웨어 교육 프로그램이다.SSAFY 제현웅 상무는 “오늘 수료하는 학생들은 모두 1년간 매일 8시간씩의 강도 높은 교육을 성실히 이수했으며 열정과 끈기를 겸비했다”라며 “삼성 청년 소프트웨어 아카데미 출신 인력들이 그동안 쌓은 역량을 발휘하며 여러 기업에서 활약하기를 기대한다”라고 말했다.삼성전자는 2018년 12월 고용노동부와 ‘삼성 청년 소프트웨어 아카데미 운영지원을 위한 업무협약(MOU)’을 체결해 교육생 선발을 위한 설명회 실시, 교육과정 홍보, 양질의 소프트웨어 교육 등 전반적인 운영부터 취업활동 지원까지 다방면으로 긴밀하게 협력하고 있다.1년간 1∙2학기 과정으로 나눠, 1학기는 기본기를 다지는 코딩 기본 교육, 2학기는 실무 중심의 프로젝트 수행 교육으로 진행되며 교육생 전원은 매달 100만원의 지원비를 받는다.1학기 기본과정은 5개월 동안 알고리즘과 코딩의 소프트웨어 기본기를 다지는 과정으로, 개인 역량에 따라 분반해 파이썬, 자바 등 프로그래밍 언어와 알고리즘, 프레임워크, 데이터베이스를 중점 학습한다.2학기 심화과정은 인공지능(AI), 사물인터넷(IoT) 등 4차 산업혁명 기술을 활용한 프로젝트를 통해 소프트웨어 실전역량을 기르는 과정으로 5개월 동안 진행된다. 1